File: android/src/main/java/com/valentingrigorean/arcgis_maps_flutter/map/LocationDisplayController.java
package com.valentingrigorean.arcgis_maps_flutter.map;

import androidx.annotation.NonNull;

import com.esri.arcgisruntime.mapping.view.Graphic;
import com.esri.arcgisruntime.mapping.view.GraphicsOverlay;
import com.esri.arcgisruntime.mapping.view.LocationDisplay;
import com.esri.arcgisruntime.mapping.view.MapView;
import com.valentingrigorean.arcgis_maps_flutter.Convert;

import java.util.ArrayList;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;

public class LocationDisplayController implements MapTouchGraphicDelegate, LocationDisplay.LocationChangedListener, MethodChannel.MethodCallHandler, LocationDisplay.AutoPanModeChangedListener, LocationDisplay.DataSourceStatusChangedListener {
    private static final String LOCATION_ATTRIBUTE = "my_location_attribute";

    private final MethodChannel channel;
    private final FlutterMapViewDelegate flutterMapViewDelegate;
    private final LocationDisplay locationDisplay;
    private final GraphicsOverlay locationGraphicsOverlay;

    private final Graphic locationGraphic;

    private LocationDisplayControllerDelegate delegate;

    private MethodChannel.Result startResult;


    private boolean trackUserLocationTap = false;

    public LocationDisplayController(MethodChannel methodChannel, FlutterMapViewDelegate flutterMapViewDelegate) {
        this.channel = methodChannel;
        this.flutterMapViewDelegate = flutterMapViewDelegate;
        this.locationDisplay = flutterMapViewDelegate.getLocationDisplay();
        this.locationGraphicsOverlay = new GraphicsOverlay();
        this.locationGraphicsOverlay.setOpacity(0);
        this.locationGraphic = new Graphic();
        this.locationGraphic.setGeometry(flutterMapViewDelegate.getLocationDisplay().getMapLocation());
        this.locationGraphic.getAttributes().put(LOCATION_ATTRIBUTE, true);
        this.locationGraphic.setSymbol(locationDisplay.getDefaultSymbol());
        this.locationGraphicsOverlay.getGraphics().add(locationGraphic);
        channel.setMethodCallHandler(this);
        locationDisplay.addDataSourceStatusChangedListener(this);
        locationDisplay.addAutoPanModeChangedListener(this);
        locationDisplay.addLocationChangedListener(this);
    }

    public void dispose() {
        locationDisplay.removeDataSourceStatusChangedListener(this);
        locationDisplay.removeAutoPanModeChangedListener(this);
        locationDisplay.removeLocationChangedListener(this);

        channel.setMethodCallHandler(null);
    }

    public void setTrackUserLocationTap(boolean trackUserLocationTap) {
        if (this.trackUserLocationTap != trackUserLocationTap) {
            this.trackUserLocationTap = trackUserLocationTap;
            if (trackUserLocationTap) {
                flutterMapViewDelegate.getGraphicsOverlays().add(locationGraphicsOverlay);
            } else {
                flutterMapViewDelegate.getGraphicsOverlays().remove(locationGraphicsOverlay);
            }
        }
    }

    public void setLocationDisplayControllerDelegate(LocationDisplayControllerDelegate delegate) {
        this.delegate = delegate;
    }

    @Override
    public boolean canConsumeTaps() {
        return trackUserLocationTap;
    }

    @Override
    public boolean didHandleGraphic(Graphic graphic) {
        final boolean result = graphic.getAttributes().containsKey(LOCATION_ATTRIBUTE);
        if (result && delegate != null) {
            delegate.onUserLocationTap();
        }
        return result;
    }

    @Override
    public void onStatusChanged(LocationDisplay.DataSourceStatusChangedEvent dataSourceStatusChangedEvent) {
        locationGraphic.setGeometry(locationDisplay.getMapLocation());
        if (startResult != null) {
            if (dataSourceStatusChangedEvent.isStarted()) {
                startResult.success(null);
            } else {
                String error = "Unknown error";
                if (dataSourceStatusChangedEvent.getError() != null) {
                    error = dataSourceStatusChangedEvent.getError().getMessage();
                }
                startResult.error("Failed to start locationDisplay", error, null);
            }
            startResult = null;
        }
    }


    @Override
    public void onAutoPanModeChanged(LocationDisplay.AutoPanModeChangedEvent autoPanModeChangedEvent) {
        channel.invokeMethod("onAutoPanModeChanged", autoPanModeChangedEvent.getAutoPanMode().ordinal());
    }

    @Override
    public void onLocationChanged(LocationDisplay.LocationChangedEvent locationChangedEvent) {
        locationGraphic.setGeometry(locationDisplay.getMapLocation());
        channel.invokeMethod("onLocationChanged", Convert.locationToJson(locationChangedEvent.getLocation()));
    }

    @Override
    public void onMethodCall(@NonNull MethodCall call, @NonNull MethodChannel.Result result) {
        switch (call.method) {
            case "getStarted":
                result.success(locationDisplay.isStarted());
                break;
            case "setAutoPanMode":
                locationDisplay.setAutoPanMode(Convert.toAutoPanMode(call.arguments));
                result.success(null);
                break;
            case "setInitialZoomScale":
                locationDisplay.setInitialZoomScale(call.arguments());
                result.success(null);
                break;
            case "setNavigationPointHeightFactor":
                locationDisplay.setNavigationPointHeightFactor(call.arguments());
                result.success(null);
                break;
            case "setWanderExtentFactor":
                locationDisplay.setWanderExtentFactor(call.arguments());
                result.success(null);
                break;
            case "getLocation":
                if (locationDisplay.getLocation() != null) {
                    result.success(Convert.locationToJson(locationDisplay.getLocation()));
                } else {
                    result.success(null);
                }
                break;
            case "getMapLocation":
                if (locationDisplay.getMapLocation() != null) {
                    result.success(Convert.geometryToJson(locationDisplay.getMapLocation()));
                } else {
                    result.success(null);
                }
                break;
            case "getHeading":
                result.success(locationDisplay.getHeading());
                break;
            case "setUseCourseSymbolOnMovement":
                locationDisplay.setUseCourseSymbolOnMovement(call.arguments());
                result.success(null);
                break;
            case "setOpacity":
                locationDisplay.setOpacity(call.arguments());
                result.success(null);
                break;
            case "setShowAccuracy":
                locationDisplay.setShowAccuracy(call.arguments());
                result.success(null);
                break;
            case "setShowLocation":
                locationDisplay.setShowLocation(call.arguments());
                break;
            case "setShowPingAnimationSymbol":
                locationDisplay.setShowPingAnimation(call.arguments());
                result.success(null);
                break;
            case "start":
                startResult = result;
                locationDisplay.startAsync();
                break;
            case "stop":
                locationDisplay.stop();
                result.success(null);
                break;
            default:
                result.notImplemented();
                break;
        }
    }

    public interface LocationDisplayControllerDelegate {
        void onUserLocationTap();
    }
}

File: android/src/main/java/com/valentingrigorean/arcgis_maps_flutter/map/LocationDisplayController.kt
package com.valentingrigorean.arcgis_maps_flutter.map

import com.esri.arcgisruntime.ArcGISRuntimeException
import com.esri.arcgisruntime.mapping.view.Graphic
import com.esri.arcgisruntime.mapping.view.GraphicsOverlay
import com.esri.arcgisruntime.mapping.view.LocationDisplay
import com.esri.arcgisruntime.mapping.view.LocationDisplay.AutoPanModeChangedEvent
import com.esri.arcgisruntime.mapping.view.LocationDisplay.AutoPanModeChangedListener
import com.esri.arcgisruntime.mapping.view.LocationDisplay.DataSourceStatusChangedEvent
import com.esri.arcgisruntime.mapping.view.LocationDisplay.DataSourceStatusChangedListener
import com.valentingrigorean.arcgis_maps_flutter.Convert
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler

class LocationDisplayController(
    private val channel: MethodChannel,
    private val flutterMapViewDelegate: FlutterMapViewDelegate
) : MapTouchGraphicDelegate, LocationDisplay.LocationChangedListener, MethodCallHandler,
    AutoPanModeChangedListener, DataSourceStatusChangedListener {
    private val locationDisplay: LocationDisplay?
    private val locationGraphicsOverlay: GraphicsOverlay
    private val locationGraphic: Graphic
    private var delegate: LocationDisplayControllerDelegate? = null
    private var startResult: MethodChannel.Result? = null
    private var trackUserLocationTap = false

    init {
        locationDisplay = flutterMapViewDelegate.locationDisplay
        locationGraphicsOverlay = GraphicsOverlay()
        locationGraphicsOverlay.opacity = 0f
        locationGraphic = Graphic()
        locationGraphic.geometry = flutterMapViewDelegate.locationDisplay.mapLocation
        locationGraphic.attributes[LOCATION_ATTRIBUTE] = true
        locationGraphic.symbol = locationDisplay!!.defaultSymbol
        locationGraphicsOverlay.graphics.add(locationGraphic)
        channel.setMethodCallHandler(this)
        locationDisplay.addDataSourceStatusChangedListener(this)
        locationDisplay.addAutoPanModeChangedListener(this)
        locationDisplay.addLocationChangedListener(this)
    }

    fun dispose() {
        locationDisplay!!.removeDataSourceStatusChangedListener(this)
        locationDisplay.removeAutoPanModeChangedListener(this)
        locationDisplay.removeLocationChangedListener(this)
        if (startResult != null) {
            startResult!!.success(null)
            startResult = null
        }
        channel.setMethodCallHandler(null)
    }

    fun setTrackUserLocationTap(trackUserLocationTap: Boolean) {
        if (this.trackUserLocationTap != trackUserLocationTap) {
            this.trackUserLocationTap = trackUserLocationTap
            if (trackUserLocationTap) {
                flutterMapViewDelegate.graphicsOverlays.add(locationGraphicsOverlay)
            } else {
                flutterMapViewDelegate.graphicsOverlays.remove(locationGraphicsOverlay)
            }
        }
    }

    fun setLocationDisplayControllerDelegate(delegate: LocationDisplayControllerDelegate?) {
        this.delegate = delegate
    }

    override fun canConsumeTaps(): Boolean {
        return trackUserLocationTap
    }

    override fun didHandleGraphic(graphic: Graphic): Boolean {
        val result = graphic.attributes.containsKey(LOCATION_ATTRIBUTE)
        if (result && delegate != null) {
            delegate!!.onUserLocationTap()
        }
        return result
    }

    override fun onStatusChanged(dataSourceStatusChangedEvent: DataSourceStatusChangedEvent) {
        try {
            locationGraphic.geometry = locationDisplay!!.mapLocation
        } catch (e: ArcGISRuntimeException) {
            //ignore
        }
        handleStatusChanged(
            dataSourceStatusChangedEvent.isStarted,
            dataSourceStatusChangedEvent.error
        )
    }

    override fun onAutoPanModeChanged(autoPanModeChangedEvent: AutoPanModeChangedEvent) {
        channel.invokeMethod("onAutoPanModeChanged", autoPanModeChangedEvent.autoPanMode.ordinal)
    }

    override fun onLocationChanged(locationChangedEvent: LocationDisplay.LocationChangedEvent) {
        try {
            locationGraphic.geometry = locationDisplay!!.mapLocation
        } catch (e: ArcGISRuntimeException) {
            //ignore
        }
        channel.invokeMethod(
            "onLocationChanged",
            Convert.Companion.locationToJson(locationChangedEvent.location)
        )
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "getStarted" -> result.success(locationDisplay!!.isStarted)
            "setAutoPanMode" -> {
                locationDisplay!!.autoPanMode =
                    Convert.Companion.toAutoPanMode(call.arguments)
                result.success(null)
            }

            "setInitialZoomScale" -> {
                locationDisplay!!.initialZoomScale = call.arguments()!!
                result.success(null)
            }

            "setNavigationPointHeightFactor" -> {
                locationDisplay!!.navigationPointHeightFactor = call.arguments()!!
                result.success(null)
            }

            "setWanderExtentFactor" -> {
                locationDisplay!!.wanderExtentFactor = call.arguments()!!
                result.success(null)
            }

            "getLocation" -> if (locationDisplay!!.location != null) {
                result.success(
                    Convert.Companion.locationToJson(
                        locationDisplay.location
                    )
                )
            } else {
                result.success(null)
            }

            "getMapLocation" -> if (locationDisplay!!.mapLocation != null) {
                result.success(
                    Convert.Companion.geometryToJson(
                        locationDisplay.mapLocation
                    )
                )
            } else {
                result.success(null)
            }

            "getHeading" -> result.success(locationDisplay!!.heading)
            "setUseCourseSymbolOnMovement" -> {
                locationDisplay!!.isUseCourseSymbolOnMovement = call.arguments()!!
                result.success(null)
            }

            "setOpacity" -> {
                locationDisplay!!.opacity = call.arguments()!!
                result.success(null)
            }

            "setShowAccuracy" -> {
                locationDisplay!!.isShowAccuracy = call.arguments()!!
                result.success(null)
            }

            "setShowLocation" -> locationDisplay!!.isShowLocation = call.arguments()!!
            "setShowPingAnimationSymbol" -> {
                locationDisplay!!.isShowPingAnimation = call.arguments()!!
                result.success(null)
            }

            "start" -> {
                startResult = result
                locationDisplay!!.startAsync()
            }

            "stop" -> {
                locationDisplay!!.stop()
                result.success(null)
            }

            else -> result.notImplemented()
        }
    }

    private fun handleStatusChanged(
        isStarted: Boolean,
        error: Throwable?
    ) {
        if (startResult != null) {
            if (isStarted) {
                startResult!!.success(null)
            } else {
                var errorMessage: String? = "Unknown error"
                if (error != null) {
                    errorMessage = error.message
                }
                startResult!!.error("Failed to start locationDisplay", errorMessage, null)
            }
            startResult = null
        }
    }

    interface LocationDisplayControllerDelegate {
        fun onUserLocationTap()
    }

    companion object {
        private const val LOCATION_ATTRIBUTE = "my_location_attribute"
    }
}